programs/marinade-finance/src/instructions/user/withdraw_stake_account.rs

```diff
@@ -0,0 +1,303 @@
+    #[account(
+        mut,
+        token::mint = msol_mint
+    )]
+    pub burn_msol_from: Box<Account<'info, TokenAccount>>,
+    #[account(mut)]
+    pub burn_msol_authority: Signer<'info>,
+
+    /// CHECK: manual account processing
+    #[account(
+        mut,
+        address = state.validator_system.validator_list.account,
+        constraint = validator_list.data.borrow().as_ref().get(0..8)
+            == Some(ValidatorSystem::DISCRIMINATOR)
+            @ MarinadeError::InvalidValidatorListDiscriminator,
+    )]
+    pub validator_list: UncheckedAccount<'info>,
+
+    /// CHECK: manual account processing
+    #[account(
+        mut,
+        address = state.stake_system.stake_list.account,
+        constraint = stake_list.data.borrow().as_ref().get(0..8)
+            == Some(StakeSystem::DISCRIMINATOR)
+            @ MarinadeError::InvalidStakeListDiscriminator,
+    )]
+    pub stake_list: UncheckedAccount<'info>,
+    /// CHECK: PDA
+    #[account(
+        seeds = [
+            &state.key().to_bytes(),
+            StakeSystem::STAKE_WITHDRAW_SEED
+        ],
+        bump = state.stake_system.stake_withdraw_bump_seed
+    )]
+    pub stake_withdraw_authority: UncheckedAccount<'info>,
+    #[account(mut)]
+    pub stake_account: Box<Account<'info, StakeAccount>>,
+
+    #[account(
+        init,
+        payer = burn_msol_authority,
+        space = std::mem::size_of::<StakeState>(),
+        owner = stake::program::ID,
+    )]
+    pub split_stake_account: Account<'info, StakeAccount>,
+
+    pub clock: Sysvar<'info, Clock>,
+    pub system_program: Program<'info, System>,
+    pub token_program: Program<'info, Token>,
+    pub stake_program: Program<'info, Stake>,
+}
+
+impl<'info> WithdrawStakeAccount<'info> {
+    pub fn process(
+        &mut self,
+        stake_index: u32,
+        validator_index: u32,
+        msol_amount: u64,
+    ) -> Result<()> {
+        require!(!self.state.paused, MarinadeError::ProgramIsPaused);
+        require!(
+            !self.state.withdraw_stake_account_enabled,
+            MarinadeError::WithdrawStakeAccountIsNotEnabled
+        );
+
+        // record  for event
+        let user_msol_balance = self.burn_msol_from.amount;
+        // save msol price source
+        let total_virtual_staked_lamports = self.state.total_virtual_staked_lamports();
+        let msol_supply = self.state.msol_supply;
+
+        check_token_source_account(
+            &self.burn_msol_from,
+            self.burn_msol_authority.key,
+            msol_amount,
+        )
+        .map_err(|e| e.with_account_name("burn_msol_from"))?;
+
+        let mut stake = self.state.stake_system.get_checked(
+            &self.stake_list.data.as_ref().borrow(),
+            stake_index,
+            self.stake_account.to_account_info().key,
+        )?;
+        let last_update_stake_delegation = stake.last_update_delegated_lamports;
+
+        // require the stake is not in emergency_unstake
+        require_neq!(
```

Comment:
require_eq! it MUST be 0 otherwise an error look here https://github.com/marinade-finance/liquid-staking-program/blob/ae1133e1f9fc611173177108464a5c7da7e8a067/programs/marinade-finance/src/instructions/crank/redelegate.rs#L144